### DISCUSSION TITLE
How to show eg variable name in violation message for XPath rules?

### DISCUSSION THREAD

**Discussion by hvbtup** [pmd/pmd#5104] [Q&A] [answered]
I'm using custom XPath rules for checking internal naming conventions etc. for PL/SQL language.
For example, a format parameter which is `IN` only should start with `pi_`, e.g. `pi_name`.

While the rule itself works well, I would like to show the variable name in the error message text (that is, in the violation description).

In #4787, there is a paragraph which makes me hope that this should work:

> XPathRule requires something to get the message for reporting violations. Currently it is getImage(). Maybe we need to extend Reportable, to provide something like a main attribute for nodes. Usually, this feature is only used for variable/field names to be present in the rule violation message.

Inside the XPath expression, I can use `@Image` to get this information.
How can I use this as a placeholder in the error message, e.g. something like

`The formal parameter $Image is input-only (IN) and thus should start with &quot;pi_&quot;.`

**Comment by jsotuyod** [comment, marked as answer]
That's correct, the `@Image` attribute (or Name, SimpleName or MethodName in that order) of the node the violation is being reported on is passed as a single argument to the message

https://github.com/pmd/pmd/blob/fc128c18ddbfa9dc4257abe0c36ba8576ce8f8d4/pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/xpath/XPathRule.java#L106-L115

You can reference this in the error message with the `{0}` placeholder.

You can see an example using an XPath rule that references a field name here

https://github.com/pmd/pmd/blob/fc128c18ddbfa9dc4257abe0c36ba8576ce8f8d4/pmd-java/src/main/resources/category/java/errorprone.xml#L386-L391

**Comment by hvbtup** [comment]
Cool!
I could not find this mentioned in the docs.
